lightning/src/ln/functional_tests.rs

```diff
@@ -8503,3 +8503,126 @@ fn test_htlc_no_detection() {
         connect_blocks(&nodes[0], ANTI_REORG_DELAY - 1, 201, true, header_201.block_hash());
         expect_payment_failed!(nodes[0], our_payment_hash, true);
 }
+
+fn do_test_onchain_htlc_settlement_after_close(broadcast_alice: bool) {
+	// If we route an HTLC, then learn the HTLC's preimage after the upstream
+	// channel has been force-closed by our counterparty, we must claim that HTLC
```

Comment:
Can we also do a test that we claim against the on-chain commitment if the channel was closed, but not on-chain when we received the HTLC update? That case should be easier, but still good to test.

Reply:
You mean to not break off-chain and go on-chain when we received the HTLC update on the forward channel ? I think we have test coverage against that so surely another interpretation.

Reply:
iiuc, added this test case.

Reply:
Have a look on L2390, `channel_monitor_network_test`, AFAICT your test extension is achieving the same as in this mentioned test we disconnect the peer with `no_connection_possible==true`, thus force-closing the channel and yet the node under test learn the preimage in-between, before connecting the counterparty commitment.

Reply:
Hm, I see what you mean. `channel_monitor_network_test` seems to cover `do_test_onchain_htlc_settlement_after_close(false, false)`, and `test_htlc_on_chain_success` covers `do_test_onchain_htlc_settlement_after_close(false, true)` (so, seems we can def get rid of those test variants).

However, for `do_test_onchain_htlc_settlement_after_close(true, false)` -- I'm not sure if we have a offchain-to-onchain claim test where the the `offchain` is settled through an `update_fulfill` message? (channel_monitor_network_test conveys the preimage directly through `claim_funds`, not through an `update_fulfill` afaict).

So, if that's the case I think it'd be OK to keep the 2 `go_onchain_before_fulfill` test variants. Lmk.

Reply:
I think it's achieving the same.

For the offchain-to-onchain as checked by `channel_monitor_network_test`, it's calling `claim_funds` which is verifying the final payment condition (MPP, ...), extracting the source then calling `claim_funds_from_hop` (L2061).

For the offchain-to-onchain as checked by `do_test_onchain_htlc_settlement_after_close(true, false)`, it's calling `internal_update_fulfill_htlc` then back `claim_funds_internal` which is itself falling back to `claim_funds_from_hop`. 

AFAICT, the code paths are different but producing the same. That said, it's not the first case of test redundancy we have and better to have a duplicate than a hole. 

Your call to keep it or not :)

Reply:
Thanks, that's a fair point, I left it in the end.